Gpu type 3

DiamonDinoia opened the pull request:
This PR implements Type3 on GPU using CUDA.

DiamonDinoia commented on src/cuda/common.cu:
```diff
@@ -164,40 +163,39 @@ void onedim_fseries_kernel_precomp(CUFINUFFT_BIGINT nf, T *f, std::complex<doubl
     f[n] = J2 * w[n] * evaluate_kernel((T)z[n], opts);    // vals & quadr wei
-    a[n] = exp((T)(2.0 * M_PI) * std::complex<T>(0.0, 1.0) * (T)(nf / 2 - z[n]) /
-               (T)nf);                                    // phase winding rates
+    a[n] = ((T)(2.0 * M_PI) * (T)(nf / 2 - z[n]) / (T)nf); // phase winding rates
```
@ahbarnett 
Together with @blackwer we figured out that we could not do the exp here and keep the polar angle.

DiamonDinoia commented on src/cuda/common.cu:
```diff
@@ -48,19 +48,18 @@ __global__ void fseries_kernel_compute(int nf1, int nf2, int nf3, T *f,
+    T x = 0.0;
     for (int n = 0; n < q; n++) {
-      x += ft[n] * 2 * (pow(cabs(at[n]), brk) * cos(brk * carg(at[n])));
+      x += ft[n] * 2 * cos(i * at[n]);
```
Together with @blackwer, finding:
@ahbarnett the magnitude of a pure imaginary number is always 1 -> pow of that is also always one so we can get rid of it entirely. The carg is using atan2 to compute the polar angle (undoing the exp from before) hence we can not do the exp in the pre-compute step and not do the atan2 here. 

This gives back the analytical formula?  Now for type 3 since the grid is not regular i (formerly brk) needs to be replaced with `k[j]`

ahbarnett commented:
Very excited you have a draft PR for t3.

Re implementing the CPU function onedim_nuft_kernel() on the GPU, this
should not be so complicated. There is simply a bunch of cosines to do. No
winding formula (which only aids the *regular* k-grid case in t1, t2).
Looking at Melody's common.cu:fseries_kernel_compute(), she uses pow, cabs,
carg... which are overly complicated, it seems. Maybe keep her cuda
blockIdx ideas but replace with
```
x += ft[n] * 2 * cos(k[i] * at[n]);
```
I think you're already realising the same thing!   Good luck! Testing
intermediate outputs may be easier than praying t3 passes unit tests in one
fell swoop... THanks, Alex


On Thu, Aug 8, 2024 at 3:38 PM Marco Barbone ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/cuda/common.cu
> <https://github.com/flatironinstitute/finufft/pull/517#discussion_r1710171741>
> :
>
> >      for (int n = 0; n < q; n++) {
> -      x += ft[n] * 2 * (pow(cabs(at[n]), brk) * cos(brk * carg(at[n])));
> +      x += ft[n] * 2 * cos(i * at[n]);
>
> Together with @blackwer <https://github.com/blackwer>, finding:
> @ahbarnett <https://github.com/ahbarnett> the magnitude of a pure
> imaginary number is always 1 -> pow of that is also always one so we can
> get rid of it entirely. The carg is using atan2 to compute the polar angle
> (undoing the exp from before) hence we can not do the exp in the
> pre-compute step and not do the atan2 here.
>
> This gives back the analytical formula? Now for type 3 since the grid is
> not regular i (formerly brk) needs to be replaced with k[j]
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flatironinstitute/finufft/pull/517#pullrequestreview-2228710227>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACNZRSW42OCRMDHQV3TAECTZQPCMRAVCNFSM6AAAAABMHD2M4KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDEMRYG4YTAMRSG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
*-------------------------------------------------------------------~^`^~._.~'
|\  Alex Barnett    Center for Computational Mathematics, Flatiron Institute
| \                 http://users.flatironinstitute.org/~ahb     646-876-5942

ahbarnett commented:
sorry, by at[n] I mean simply z[n], as per the CPU code. I don't understand
Melody's use of the at array (it may follow my OMP implementation using aj
in the CPU code). Point is, just plainly eval the cosines in the first GPU
version; then we see if any HPC needed there later... :)

On Fri, Aug 9, 2024 at 8:40 PM Alex Barnett ***@***.***>
wrote:

> Very excited you have a draft PR for t3.
>
> Re implementing the CPU function onedim_nuft_kernel() on the GPU, this
> should not be so complicated. There is simply a bunch of cosines to do. No
> winding formula (which only aids the *regular* k-grid case in t1, t2).
> Looking at Melody's common.cu:fseries_kernel_compute(), she uses pow,
> cabs, carg... which are overly complicated, it seems. Maybe keep her cuda
> blockIdx ideas but replace with
> ```
> x += ft[n] * 2 * cos(k[i] * at[n]);
> ```
> I think you're already realising the same thing!   Good luck! Testing
> intermediate outputs may be easier than praying t3 passes unit tests in one
> fell swoop... THanks, Alex
>
>
> On Thu, Aug 8, 2024 at 3:38 PM Marco Barbone ***@***.***>
> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In src/cuda/common.cu
>> <https://github.com/flatironinstitute/finufft/pull/517#discussion_r1710171741>
>> :
>>
>> >      for (int n = 0; n < q; n++) {
>> -      x += ft[n] * 2 * (pow(cabs(at[n]), brk) * cos(brk * carg(at[n])));
>> +      x += ft[n] * 2 * cos(i * at[n]);
>>
>> Together with @blackwer <https://github.com/blackwer>, finding:
>> @ahbarnett <https://github.com/ahbarnett> the magnitude of a pure
>> imaginary number is always 1 -> pow of that is also always one so we can
>> get rid of it entirely. The carg is using atan2 to compute the polar angle
>> (undoing the exp from before) hence we can not do the exp in the
>> pre-compute step and not do the atan2 here.
>>
>> This gives back the analytical formula? Now for type 3 since the grid is
>> not regular i (formerly brk) needs to be replaced with k[j]
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/flatironinstitute/finufft/pull/517#pullrequestreview-2228710227>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACNZRSW42OCRMDHQV3TAECTZQPCMRAVCNFSM6AAAAABMHD2M4KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDEMRYG4YTAMRSG4>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
>
> *-------------------------------------------------------------------~^`^~._.~'
> |\  Alex Barnett    Center for Computational Mathematics, Flatiron
> Institute
> | \                 http://users.flatironinstitute.org/~ahb
>  646-876-5942
>
>

-- 
*-------------------------------------------------------------------~^`^~._.~'
|\  Alex Barnett    Center for Computational Mathematics, Flatiron Institute
| \                 http://users.flatironinstitute.org/~ahb     646-876-5942

DiamonDinoia commented:
This first version now works. However there it seems to be an issue with oversampling factor 1.25 and high tol = 1e-8, disabling horner or using upsampfact = 2.00 fixes it. It might be the coefficients, but more investigation is needed.

lu1and10 commented on src/finufft.cpp:
```diff
@@ -754,13 +755,12 @@ int FINUFFT_MAKEPLAN(int type, int dim, BIGINT *n_modes, int iflag, int ntrans,
       // Unfortunately fftw_planner_nthreads wasn't introduced until fftw 3.3.9, and
       // there isn't a convenient mechanism to probe the version
-      FFTW_PLAN_TH(nthr_fft);
+      // there is fftw_version which returns a string, but that's not compile time
```
Seems PR  #526 is trying to address this part.

DiamonDinoia replied:
I need to revert this.

lu1and10 commented on examples/CMakeLists.txt:
Changes in this file may have some overlap with PR  #535. @jkrimmer

DiamonDinoia commented:
Changelog:
* Support for type 3 in 1D, 2D, and 3D in the GPU library cufinufft (PR #517).
* Removed the CPU fseries computation (only used for benchmark no longer needed).
* Added complex arithmetic support for cuda_complex type
* Added tests for type 3 in 1D, 2D, and 3D and cuda_complex arithmetic
* Minor fixes on the GPU code:
    - removed memory leaks in case of errors
    - renamed maxbatchsize to batchsize

DiamonDinoia commented:
The test failure is not an issue. I will disable that test later.

ahbarnett commented on src/cuda/common.cu:
```diff
@@ -26,15 +26,18 @@ using std::max;
+  T *at = a + threadIdx.y * MAX_NQUAD;
+  T *ft = f + threadIdx.y * MAX_NQUAD;
   T *oarr;
+  // standard parallelism pattern in cuda. using a 2D grid, this allows to leverage more
```
good comments... otherwise no reader would figure out what the threadIdx.y is doing!

ahbarnett commented on src/cuda/common.cu:
```diff
@@ -26,15 +26,18 @@ using std::max;
-__global__ void fseries_kernel_compute(int nf1, int nf2, int nf3, T *f,
-                                       cuDoubleComplex *a, T *fwkerhalf1, T *fwkerhalf2,
-                                       T *fwkerhalf3, int ns) {
+__global__ void fseries_kernel_compute(int nf1, int nf2, int nf3, T *f, T *a,
```
Could you comment that this is the equispaced frequency case, used by type 1 & 2, matching onedim_fseries_kernel in CPU code?   And the other fseries_kernel_compute is arbitrary freqs, used by type 3 only?   Otherwise the overlaoding is confusing. thnx.

ahbarnett commented on src/cuda/common.cu:
add that implements onedim_nuft_kernel in CPU code.

ahbarnett commented on src/cuda/common.cu:
In this GPU code there actually is no phase winding used (that was a CPU speedup for equispaced case for type 1 & 2).   I might be tempted to tidy up this section, writing two separate funcs (with names as in CPU code), each of which includes its own setup of phase vector "a"... maybe this can wait til after merge. Depends how you feel.

DiamonDinoia replied:
I think we should wait. This might be tidied up combining also flipwind

ahbarnett commented on src/cuda/common.cu:
really, a is always a vector of scaled quadrature nodes; the scaling happens to be different for type 1,2 (equispaced) vs type 3 (general kx,ky,kz).  Really having them both called "a" is confusing... see previous review.

ahbarnett commented on src/finufft.cpp:
```diff
@@ -726,6 +725,7 @@ int FINUFFT_MAKEPLAN(int type, int dim, BIGINT *n_modes, int iflag, int ntrans,
       fprintf(stderr,
               "[%s] fwBatch would be bigger than MAX_NF, not attempting malloc!\n",
               __func__);
+      // FIXME: this error causes memory leaks. We should free phiHat1, phiHat2, phiHat3
```
do the FIXME ;)

DiamonDinoia replied:
Let's leave it in a separate PR. Otherwise in case of issues reverting this also removes an unrelated fix.

ahbarnett commented on test/cuda/CMakeLists.txt:
```diff
@@ -71,17 +92,30 @@ function(add_tests PREC REQ_TOL CHECK_TOL UPSAMP)
-
-add_test(NAME cufinufft_public_api COMMAND public_api_test)
-add_test(NAME cufinufft_makeplan COMMAND test_makeplan)
+add_tests(float 1e-5 2e-4 0.f)
```
I'm confused how the UPSAMP=0.f string appends to the executable here? Does "f" mean float, in which case, should all the add_tests(float *) be with "f" ? If it's correct, please add comment.

ahbarnett commented on test/cuda/cufinufft1d_test.cu:
```diff
@@ -50,6 +51,16 @@ int run_test(int method, int type, int N1, int M, T tol, T checktol, int iflag,
+    }
+    s.resize(N1);
+    for (int i = 0; i < N1; i++) {
+      s[i] = M_PI * randm11();
```
In test/finufft1d_test.cpp the "s" array is scaled by S = N1/2.  I suggest you do this here too, to make the space-freq product grow (hence the FFT size) as the number of modes. This matches CPU testers. Also for 2d, 3d, below.

ahbarnett commented on test/cuda/cufinufft1d_test.cu:
```diff
@@ -174,6 +191,16 @@ int run_test(int method, int type, int N1, int M, T tol, T checktol, int iflag,
+      Ft += c[j] * exp(J * (x[j] * s[jt]));
+    }
+    rel_error = abs(Ft - fk[jt]) / infnorm(N1, (std::complex<T> *)fk.data());
+    printf("[gpu   ] one mode: rel err in F[%d] is %.3g\n", jt, rel_error);
```
Line 215 below: (1,2) -> (1,2,3) in usage string.

ahbarnett commented on test/cuda/cufinufft2d_test.cu:
```diff
@@ -49,6 +49,19 @@ int run_test(int method, int type, int N1, int N2, int M, T tol, T checktol, int
+    s.resize(N1 * N2);
+    t.resize(N1 * N2);
+    for (int i = 0; i < N1 * N2; i++) {
+      s[i] = M_PI * randm11();
```
see prev review on N1, N2 scaling of the s,t test freq pts.

ahbarnett commented on test/cuda/cufinufft3d_test.cu:
```diff
@@ -51,6 +51,22 @@ int run_test(int method, int type, int N1, int N2, int N3, int M, T tol, T check
+    t.resize(N1 * N2 * N3);
+    u.resize(N1 * N2 * N3);
+    for (int i = 0; i < N1 * N2 * N3; i++) {
+      s[i] = M_PI * randm11();
```
ditto scaling above.

ahbarnett commented on test/cuda/cufinufft_math_test.cu:
```diff
@@ -0,0 +1,130 @@
+
+// Helper function to compare cuComplex with std::complex<T> using 1 - ratio as error
+template<typename T>
+bool compareComplex(const cuda_complex<T> &a, const std::complex<T> &b,
```
As we discussed today, this is too harsh a test, since when a complex number "lands" near the real axis, its imag part (by itself) may have high relative error, and that's ok. Instead of separately testing Re and Im rel errors, use cabs(a-b)/cabs(a)  < epsilon.    Ask if still doesn't make sense.

ahbarnett reviewed (changes requested):
Great job! A big one.  Just a few things to tidy up, should be at most 1-2 hrs if you don't address the the phase-winding tidy-up.

[However, have a look at removing the whole phase-winding and "a" stuff (it is way more confusing that it needs be, due to old code left by Melody and Robert), which I could have a stab at after merge, anyway, unless you want to. It might add 0.5-1 days work for you. Sit down and write out the math, or go back to the CPU code sans phase-winding, just using plain cos().  If it's not clear, we can together after Sept 3.]

janden reviewed (changes requested):
Looks good on my end. Just a few questions.

DiamonDinoia commented on include/cufinufft/impl.h:
```diff
@@ -76,205 +83,252 @@ int cufinufft_makeplan_impl(int type, int dim, int *nmodes, int iflag, int ntran
+    using namespace cufinufft::memtransfer;
+    switch (d_plan->dim) {
+    case 1: {
+      if ((ier = allocgpumem1d_plan<T>(d_plan))) goto finalize;
```
`goto` were already used in the code. I left it because it is a good way to do error management. `goto` can only be use for this particular use case and they are quite clear.

DiamonDinoia commented on include/cufinufft/types.h:
```diff
@@ -37,7 +37,7 @@ template<typename T> struct cufinufft_plan_t {
   CUFINUFFT_BIGINT mu;
   int ntransf;
-  int maxbatchsize;
+  int batchsize;
```
In the opts yes it is maxbatchsize. However in the plan is just batchsize as the actual batchsize can be smaller than maxbatchsize

DiamonDinoia commented on include/cufinufft/utils.h:
```diff
@@ -146,6 +137,76 @@ static __forceinline__ __device__ void atomicAddComplexGlobal(
+  return std::make_tuple(w, c);
+}
+
+template<typename T>
```
Because then cufinufft will have to build and link against finufft. I do not think is worth bringing finufft as a dependency of cufinufft as it is. Once finufft is de-macroized it will be worth grouping common functionalities in a library that is imported by both.

DiamonDinoia commented on src/cuda/3d/spread3d_wrapper.cu:
```diff
@@ -5,6 +5,7 @@
 #include <cufinufft/contrib/helper_cuda.h>
 #include <thrust/device_ptr.h>
 #include <thrust/scan.h>
+#include <thrust/sort.h>
```
merge sort in 1D

DiamonDinoia commented:
Integrated most of the requested changes. Commented on the PR when not applicable. 
On the timing, maybe worth relying on the nvidia profiler as it provides a nice output and is shipped with cuda. Otherwise I will think about it, as one has to use the CUDA event API to measure the time. 

the fseries should be reworked in a separate PR as it can be combined with the flipwind changes.
